core/style.js

```diff
@@ -1662,7 +1662,8 @@ CKEDITOR.STYLE_OBJECT = 3;
 
 		// Avoid ID duplication.
 		if ( targetDocument.getCustomData( 'doc_processing_style' ) && el.hasAttribute( 'id' ) )
-			el.removeAttribute( 'id' );
+			if ( targetDocument.getById(el.getAttribute('id')) != null )
```

Comment:
This is not a proper fix for this. Actually the problem originates from bad `doc_processing_style` custom data management.

If you trace the code changes, you'll notice that it was moved from `setupElement` function to `applyStyleOnSelection` - and this is exactly when the bug was introduced.

This attribute was used to, sort of, mark a single transaction - but in case of editing an existing link this function wasn't even called, thus it didn't clear this "switch".

All that is needed here is to restore it and it should work correctly.

Reply:
Also the code style is inconsistent, see spacing next to parenthesis.